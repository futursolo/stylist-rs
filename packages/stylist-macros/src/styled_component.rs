// This file is borrowed from yew-macro/src/function_component.rs
use std::iter::FromIterator;

use proc_macro2::{Group, Span, TokenStream, TokenTree};
use quote::{quote, ToTokens};
use syn::parse::{Parse, ParseStream};
use syn::parse_macro_input;
use syn::{Ident, Item, ItemFn};

#[derive(Debug)]
pub struct StyledComponent {
    func: ItemFn,
}

impl Parse for StyledComponent {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let parsed: Item = input.parse()?;

        match parsed {
            Item::Fn(func) => Ok(Self { func }),
            item => Err(syn::Error::new_spanned(
                item,
                "`styled_component` attribute can only be applied to functions",
            )),
        }
    }
}

#[derive(Debug)]
pub struct StyledComponentName {
    component_name: Ident,
}

impl Parse for StyledComponentName {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        if input.is_empty() {
            return Err(input.error("expected identifier for the component"));
        }

        let component_name = input.parse()?;

        Ok(Self { component_name })
    }
}

#[derive(Debug, PartialEq)]
enum State {
    NotFound,
    Ident,
    Excl,
}

// affix a .with_manager(__stylist_style_manager__) when after `css!` (or other macro name).
pub fn affix_manager(macro_name: &str, input: TokenStream, mgr_ident: Ident) -> TokenStream {
    let tokens = input.into_iter();

    let mut completed: Vec<TokenTree> = Vec::new();

    let mut state = State::NotFound;

    for token in tokens {
        match token {
            TokenTree::Ident(ref m) => {
                if state == State::NotFound && *m == macro_name {
                    state = State::Ident;
                } else {
                    state = State::NotFound;
                }

                completed.push(token);
            }

            TokenTree::Punct(ref m) => {
                if state == State::Ident && m.as_char() == '!' {
                    state = State::Excl;
                } else {
                    state = State::NotFound;
                }

                completed.push(token);
            }
            TokenTree::Literal(_) => {
                state = State::NotFound;
                completed.push(token);
            }
            TokenTree::Group(ref m) => {
                let content = affix_manager(macro_name, m.stream(), mgr_ident.clone());
                let group_tokens = Group::new(m.delimiter(), content).to_token_stream();

                completed.extend(group_tokens);

                if state == State::Excl {
                    completed.extend(quote! {
                        .with_manager({
                            #[allow(clippy::redundant_clone)]
                            #mgr_ident.clone()
                        })
                    });
                }

                state = State::NotFound;
            }
        }
    }

    TokenStream::from_iter(completed)
}

pub fn styled_component_impl(
    name: StyledComponentName,
    component: StyledComponent,
) -> syn::Result<TokenStream> {
    let StyledComponentName { component_name } = name;

    let StyledComponent { func } = component;

    let ItemFn {
        attrs,
        vis,
        sig,
        block,
    } = func;

    // Turn into token stream, so it's easier to process.
    let block_tokens = block.to_token_stream();

    let mgr_ident = Ident::new("__stylist_style_manager__", Span::mixed_site());
    let justified_tokens = affix_manager("css", block_tokens, mgr_ident.clone());

    let quoted = quote! {
        #[::yew::functional::function_component(#component_name)]
        #(#attrs)*
        #vis #sig {
            let #mgr_ident = ::yew::functional::use_context::<::stylist::manager::StyleManager>().unwrap_or_default();
            #[allow(unused_imports)]
            use ::stylist::css;

            #justified_tokens
        }
    };

    Ok(quoted)
}

pub fn macro_fn(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let item = parse_macro_input!(item as StyledComponent);
    let attr = parse_macro_input!(attr as StyledComponentName);

    styled_component_impl(attr, item)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}
