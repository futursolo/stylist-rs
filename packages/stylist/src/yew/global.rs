use yew::prelude::*;

use crate::manager::StyleManager;
use crate::{GlobalStyle, StyleSource};
use stylist_core::ResultDisplay;

/// The properties for the [`Global`] Component, please see its documentation for usage.
#[derive(Properties, Clone, Debug, PartialEq)]
pub struct GlobalProps {
    pub css: StyleSource,
}

/// A Global Style that will be applied to `<html />` tag, inspired by [emotion](https://emotion.sh).
///
/// The `css` attribute accepts a value of any type that implements
/// [`IntoPropValue<StyleSource>`](yew::html::IntoPropValue). If you use the `parser`
/// feature, this panics when supplying a string that fails to parse.
///
/// The style will be applied via the [`:root`](https://developer.mozilla.org/en-US/docs/Web/CSS/:root)
/// pseudo-class, which has higher specificity than "simply" using the `html` selector.
///
/// # Example:
///
/// ```
/// use stylist::css;
/// use stylist::yew::Global;
/// use yew::prelude::*;
///
/// #[function_component]
/// fn App() -> Html {
///     html! {
///         <>
///             <Global css={css!("color: red;")} />
///             <div>{"Hello World!"}</div>
///         </>
///     }
/// }
/// ```
#[function_component(Global)]
pub fn global(props: &GlobalProps) -> Html {
    let GlobalProps { css } = props.clone();
    let mgr = use_context::<StyleManager>().unwrap_or_default();

    struct GlobalStyleGuard {
        inner: GlobalStyle,
    }

    impl Drop for GlobalStyleGuard {
        fn drop(&mut self) {
            self.inner.unregister();
        }
    }

    use_memo((mgr, css), move |(manager, css)| GlobalStyleGuard {
        inner: GlobalStyle::new_with_manager(css.clone(), manager)
            .expect_display("Failed to create style."),
    });

    Html::default()
}
