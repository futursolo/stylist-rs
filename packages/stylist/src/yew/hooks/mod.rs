#[cfg_attr(documenting, doc(cfg(feature = "yew_use_media_query")))]
#[cfg(feature = "yew_use_media_query")]
mod use_media_query;

#[cfg_attr(documenting, doc(cfg(feature = "yew_use_media_query")))]
#[cfg(feature = "yew_use_media_query")]
pub use use_media_query::use_media_query;
