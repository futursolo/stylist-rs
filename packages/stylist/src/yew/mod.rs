//! This module contains yew specific features.

use yew::html::Classes;
use yew::html::IntoPropValue;

/// A procedural macro to style Yew component.
///
/// # Example:
///
/// ```rust
/// use std::borrow::Cow;
///
/// use yew::prelude::*;
/// use stylist::yew::styled_component;
///
/// #[styled_component(MyStyledComponent)]
/// fn my_styled_component() -> Html {
///     html! {<div class={css!("color: red;")}>{"Hello World!"}</div>}
/// }
/// ```
///
/// # Note:
///
/// You don't need to import [`css!`](crate::css) inside of a `styled_component`.
///
/// This macro imports a special version of [`css!`](crate::css) macro that is aware of the current style manager.
#[cfg_attr(documenting, doc(cfg(feature = "macros")))]
#[cfg(feature = "macros")]
pub use stylist_macros::styled_component;

use crate::ast::Sheet;
use crate::manager::StyleManager;
use crate::{Style, StyleSource};

use yew::html::ImplicitClone;

impl ImplicitClone for StyleManager {}

mod global;
mod hooks;
mod provider;

pub use global::{Global, GlobalProps};
pub use provider::{ManagerProvider, ManagerProviderProps};

pub use hooks::*;

impl From<Style> for Classes {
    fn from(style: Style) -> Self {
        let mut classes = Self::new();
        classes.push(style.get_class_name().to_string());
        classes
    }
}

impl From<StyleSource<'_>> for Classes {
    fn from(style_src: StyleSource<'_>) -> Self {
        let mut classes = Self::new();
        classes.push(style_src.to_style().get_class_name().to_string());
        classes
    }
}

impl IntoPropValue<StyleSource<'static>> for Sheet {
    fn into_prop_value(self) -> StyleSource<'static> {
        self.into()
    }
}

#[cfg_attr(documenting, doc(cfg(feature = "parser")))]
#[cfg(feature = "parser")]
mod feat_parser {
    use std::borrow::Cow;

    use super::*;

    impl IntoPropValue<StyleSource<'static>> for String {
        fn into_prop_value(self) -> StyleSource<'static> {
            self.into()
        }
    }

    impl<'a> IntoPropValue<StyleSource<'a>> for &'a str {
        fn into_prop_value(self) -> StyleSource<'a> {
            self.into()
        }
    }

    impl<'a> IntoPropValue<StyleSource<'a>> for Cow<'a, str> {
        fn into_prop_value(self) -> StyleSource<'a> {
            self.into()
        }
    }
}
