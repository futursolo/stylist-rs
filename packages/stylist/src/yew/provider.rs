use yew::prelude::*;

use crate::manager::StyleManager;

/// The properties for [`ManagerProvider`] Component, please see its documentation for usage.
#[derive(Properties, Clone, Debug, PartialEq)]
pub struct ManagerProviderProps {
    pub manager: StyleManager,
    #[prop_or_default]
    pub children: Children,
}

/// A Context Provider to provide a custom [`StyleManager`] to child components.
///
/// # Example:
///
/// ```
/// use stylist::manager::StyleManager;
/// use stylist::yew::ManagerProvider;
/// use yew::prelude::*;
///
/// #[function_component(App)]
/// fn app() -> Html {
///     let mgr = use_state(|| {
///         StyleManager::builder()
///             .prefix("my-styles".into())
///             .build()
///             .unwrap()
///     });
///
///     let children = Html::default();
///     html! {
///         <ManagerProvider manager={(*mgr).clone()}>
///             {children}
///         </ManagerProvider>
///     }
/// }
/// ```
#[function_component(ManagerProvider)]
pub fn manager_provider(props: &ManagerProviderProps) -> HtmlResult {
    let ManagerProviderProps { manager, children } = props.clone();

    #[cfg(any(feature = "ssr", feature = "hydration"))]
    {
        use crate::manager::StyleData;

        let _manager = manager.clone();
        let _style_data = use_transitive_state!(
            move |_| -> StyleData {
                _manager
                    .style_data()
                    .map(|m| m.clone())
                    .unwrap_or_else(StyleData::new)
            },
            ()
        )?;

        #[cfg(feature = "hydration")]
        {
            // We must load the styles immediately before child components are rendered.
            let manager = manager.clone();
            use_memo(
                move |manager| {
                    if let Some(m) = _style_data {
                        manager.load_style_data(m.as_ref());
                    }
                },
                manager,
            );
        }
    }

    Ok(html! {
        <ContextProvider<StyleManager> context={manager}>
            {children}
        </ContextProvider<StyleManager>>
    })
}
