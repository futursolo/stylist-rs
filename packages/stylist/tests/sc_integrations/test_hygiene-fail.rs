use stylist::yew::styled_component;
use yew::prelude::*;

#[styled_component(App)]
pub fn app() -> Html {
    let mgr = __stylist_style_manager__.clone();

    Html::default()
}

fn main() {}
