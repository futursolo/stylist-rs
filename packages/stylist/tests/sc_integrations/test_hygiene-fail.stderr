error[E0425]: cannot find value `__stylist_style_manager__` in this scope
 --> $DIR/test_hygiene-fail.rs:6:15
  |
6 |     let mgr = __stylist_style_manager__.clone();
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope
